File: src/test/java/com/robohorse/robopojogenerator/generator/processors/ClassTemplateHelperTest.java
package com.robohorse.robopojogenerator.generator.processors;

import com.robohorse.robopojogenerator.generator.consts.ClassTemplate;
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper;
import com.robohorse.robopojogenerator.generator.utils.ClassTemplateHelper;
import com.robohorse.robopojogenerator.models.FieldModel;
import org.junit.Before;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.when;

/**
 * Created by vadim on 05.10.16.
 */
public class ClassTemplateHelperTest {
    @InjectMocks
    ClassTemplateHelper classTemplateHelper;

    @Mock
    ClassGenerateHelper classGenerateHelper;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void testCreateSetter() throws Exception {
        final String field = "item";
        final String type = "String";
        final String fieldUpper = "Item";
        when(classGenerateHelper.upperCaseFirst(field))
                .thenReturn(fieldUpper);
        when(classGenerateHelper.lowerCaseFirst(field))
                .thenReturn(field);
        when(classGenerateHelper.formatClassField(field))
                .thenReturn(field);

        final String target = ClassTemplate.TAB + "public void set" + fieldUpper + "(" + type + " " + field + "){"
                + ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + ClassTemplate.TAB + "this." + field + " = " + field + ";"
                + ClassTemplate.NEW_LINE
                + ClassTemplate.TAB + "}"
                + ClassTemplate.NEW_LINE;
        assertEquals(target, classTemplateHelper.createSetter(field, type));
    }

    @Test
    public void testCreateGetter() throws Exception {
        final String field = "item";
        final String type = "String";
        final String fieldUpper = "Item";
        when(classGenerateHelper.upperCaseFirst(field))
                .thenReturn(fieldUpper);
        when(classGenerateHelper.lowerCaseFirst(field))
                .thenReturn(field);
        when(classGenerateHelper.formatClassField(field))
                .thenReturn(field);

        final String target = ClassTemplate.TAB + "public " + type + " get" + fieldUpper + "(){"
                + ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + ClassTemplate.TAB + "return " + field + ";"
                + ClassTemplate.NEW_LINE
                + ClassTemplate.TAB + "}"
                + ClassTemplate.NEW_LINE;
        assertEquals(target, classTemplateHelper.createGetter(field, type));
    }

    @Test
    public void testCreateGetterBoolean() throws Exception {
        final String field = "item";
        final String type = "boolean";
        final String fieldUpper = "Item";
        when(classGenerateHelper.upperCaseFirst(field))
                .thenReturn(fieldUpper);
        when(classGenerateHelper.lowerCaseFirst(field))
                .thenReturn(field);
        when(classGenerateHelper.formatClassField(field))
                .thenReturn(field);

        final String target = ClassTemplate.TAB + "public boolean is" + fieldUpper + "(){"
                + ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + ClassTemplate.TAB + "return " + field + ";"
                + ClassTemplate.NEW_LINE
                + ClassTemplate.TAB + "}"
                + ClassTemplate.NEW_LINE;
        assertEquals(target, classTemplateHelper.createGetter(field, type));
    }

    @Test
    public void testCreateField() throws Exception {
        final String field = "item";
        final String type = "boolean";
        final String target = ClassTemplate.TAB + "private " + type + " " + field + ";"
                + ClassTemplate.NEW_LINE;
        when(classGenerateHelper.formatClassField(field))
                .thenReturn(field);
        assertEquals(target, classTemplateHelper.createFiled(
                new FieldModel.Builder()
                        .setClassType(type)
                        .setFieldNameFormatted(field)
                        .setFieldName(field)
                        .build()
        ));
    }

    @Test
    public void testCreateFieldGenerated() throws Exception {
        final String field = "item";
        final String type = "boolean";
        final String annotation = "@JsonField";
        final String target = ClassTemplate.NEW_LINE + ClassTemplate.TAB + annotation + ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + "private " + type + " " + field + ";"
                + ClassTemplate.NEW_LINE;
        when(classGenerateHelper.formatClassField(field))
                .thenReturn(field);
        assertEquals(target, classTemplateHelper.createFiled(
                new FieldModel.Builder()
                        .setClassType(type)
                        .setAnnotation(annotation)
                        .setFieldNameFormatted(field)
                        .setFieldName(field)
                        .build()
        ));
    }
}

File: generator/src/test/kotlin/com/robohorse/robopojogenerator/generator/processing/ClassTemplateHelperTest.kt
package com.robohorse.robopojogenerator.generator.processing

import com.robohorse.robopojogenerator.generator.common.templates.ClassTemplate
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper
import com.robohorse.robopojogenerator.generator.utils.ClassTemplateHelper
import com.robohorse.robopojogenerator.models.FieldModel
import com.robohorse.robopojogenerator.models.Visibility
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.RelaxedMockK
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

internal class ClassTemplateHelperTest {
    @RelaxedMockK
    lateinit var classGenerateHelper: ClassGenerateHelper

    @InjectMockKs
    lateinit var classTemplateHelper: ClassTemplateHelper

    @BeforeTest
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    fun testCreateSetter() {
        val field = "item"
        val type = "String"
        val fieldUpper = "Item"
        every { classGenerateHelper.upperCaseFirst(field) }.returns(fieldUpper)
        every { classGenerateHelper.lowerCaseFirst(field) }.returns(field)
        every { classGenerateHelper.formatClassField(field) }.returns(field)
        val target = (
            ClassTemplate.TAB + "public void set" + fieldUpper + "(" + type + " " + field + "){" +
                ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + ClassTemplate.TAB + "this." + field + " = " + field + ";" +
                ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + "}" +
                ClassTemplate.NEW_LINE
            )
        assertEquals(target, classTemplateHelper.createSetter(field, type))
    }

    @Test
    fun testCreateGetter() {
        val field = "item"
        val type = "String"
        val fieldUpper = "Item"
        every { classGenerateHelper.upperCaseFirst(field) }.returns(fieldUpper)
        every { classGenerateHelper.lowerCaseFirst(field) }.returns(field)
        every { classGenerateHelper.formatClassField(field) }.returns(field)
        val target = (
            ClassTemplate.TAB + "public " + type + " get" + fieldUpper + "(){" +
                ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + ClassTemplate.TAB + "return " + field + ";" +
                ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + "}" +
                ClassTemplate.NEW_LINE
            )
        assertEquals(target, classTemplateHelper.createGetter(field, type))
    }

    @Test
    fun testCreateGetterBoolean() {
        val field = "item"
        val type = "boolean"
        val fieldUpper = "Item"
        every { classGenerateHelper.upperCaseFirst(field) }.returns(fieldUpper)
        every { classGenerateHelper.lowerCaseFirst(field) }.returns(field)
        every { classGenerateHelper.formatClassField(field) }.returns(field)
        val target = (
            ClassTemplate.TAB + "public boolean is" + fieldUpper + "(){" +
                ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + ClassTemplate.TAB + "return " + field + ";" +
                ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + "}" +
                ClassTemplate.NEW_LINE
            )
        assertEquals(target, classTemplateHelper.createGetter(field, type))
    }

    @Test
    fun testCreateField() {
        val field = "item"
        val type = "boolean"
        val target = (
            ClassTemplate.TAB + "private " + type + " " + field + ";" +
                ClassTemplate.NEW_LINE
            )
        every { classGenerateHelper.formatClassField(field) }.returns(field)
        assertEquals(
            target,
            classTemplateHelper.createField(
                FieldModel(
                    classType = type,
                    fieldNameFormatted = field,
                    fieldName = field,
                    visibility = Visibility.PRIVATE
                )
            )
        )
    }

    @Test
    fun testCreateFieldGenerated() {
        val field = "item"
        val type = "boolean"
        val annotation = "@JsonField"
        val target = (
            ClassTemplate.NEW_LINE + ClassTemplate.TAB + annotation + ClassTemplate.NEW_LINE +
                ClassTemplate.TAB + "private " + type + " " + field + ";" +
                ClassTemplate.NEW_LINE
            )
        every { classGenerateHelper.formatClassField(field) }.returns(field)
        assertEquals(
            target,
            classTemplateHelper.createField(
                FieldModel(
                    classType = type,
                    fieldName = field,
                    fieldNameFormatted = field,
                    annotation = annotation,
                    visibility = Visibility.PRIVATE
                )
            )
        )
    }

    @Test
    fun testCreateFieldWithoutVisibility() {
        val field = "item"
        val type = "boolean"
        val target = (ClassTemplate.TAB + type + " " + field + ";" + ClassTemplate.NEW_LINE)
        every { classGenerateHelper.formatClassField(field) }.returns(field)
        assertEquals(
            target,
            classTemplateHelper.createField(
                FieldModel(
                    classType = type,
                    fieldNameFormatted = field,
                    fieldName = field,
                    visibility = Visibility.NONE
                )
            )
        )
    }
}
